service/worker/deployment/deployment_util.go

```diff
@@ -38,80 +39,80 @@ import (
-// workflowID which are used in our deployment workflows
-func GenerateDeploymentWorkflowID(deploymentName string, buildID string) string {
-	return DeploymentWorkflowIDPrefix + DeploymentWorkflowIDDelimeter + deploymentName + DeploymentWorkflowIDDelimeter + buildID
+type DeploymentClient interface {
+	RegisterWorker(
+		ctx context.Context,
+		taskQueueName string,
+		taskQueueType enumspb.TaskQueueType,
+		pollTimestamp time.Time,
+	) error
 }
 
-// ValidateDeploymentWfParams is a helper that verifies if the fields used for generating
-// deployment related workflowID's are valid
-func ValidateDeploymentWfParams(fieldName string, field string, maxIDLengthLimit int) error {
-	// Length checks
-	if field == "" {
-		return serviceerror.NewInvalidArgument(fmt.Sprintf("%v cannot be empty", fieldName))
-	}
-
-	// Length of each field should be: (MaxIDLengthLimit - prefix and delimeter length) / 2
-	if len(field) > (maxIDLengthLimit-DeploymentWorkflowIDInitialSize)/2 {
-		return serviceerror.NewInvalidArgument(fmt.Sprintf("size of %v larger than the maximum allowed", fieldName))
-	}
+// implements DeploymentClient
+type DeploymentWorkflowClient struct {
+	namespace     *namespace.Namespace
+	deployment    *commonpb.WorkerDeployment
+	historyClient resource.HistoryClient
+}
 
-	// Invalid character check - Cannot contain the reserved delimeter ("|") since its being used for building deployment workflowID
-	if strings.Contains(field, DeploymentWorkflowIDDelimeter) {
-		return serviceerror.NewInvalidArgument(fmt.Sprintf("%v cannot contain reserved prefix %v", fieldName, DeploymentWorkflowIDDelimeter))
+func NewDeploymentWorkflowClient(
+	namespace *namespace.Namespace,
+	deployment *commonpb.WorkerDeployment,
+	historyClient resource.HistoryClient,
+) *DeploymentWorkflowClient {
+	return &DeploymentWorkflowClient{
+		namespace:     namespace,
+		deployment:    deployment,
+		historyClient: historyClient,
 	}
-
-	// UTF-8 check
-	return common.ValidateUTF8String(fieldName, field)
 }
 
-// StartAndUpdateDeploymentWorkflow is a helper that validates the input and update workflow parameters
-// and uses a multi-operation to start and update a deployment workflow.
-func StartAndUpdateDeploymentWorkflow(
+func (d *DeploymentWorkflowClient) RegisterTaskQueueWorker(
 	ctx context.Context,
-	workflowArgs *deployspb.DeploymentWorkflowArgs,
-	updateArgs *deployspb.RegisterWorkerInDeploymentArgs,
+	taskQueueName string,
+	taskQueueType enumspb.TaskQueueType,
+	pollTimestamp *timestamppb.Timestamp,
 	maxIDLengthLimit int,
-	historyClient resource.HistoryClient,
-) (*historyservice.ExecuteMultiOperationResponse, error) {
+) error {
 	sa := &commonpb.SearchAttributes{}
 	searchattribute.AddSearchAttribute(&sa, searchattribute.TemporalNamespaceDivision, payload.EncodeString(DeploymentNamespaceDivision))
 
+	// escaping the reserved workflow delimiter (|) from the inputs, if present
+	d.deployment.DeploymentName = d.EscapeChar(d.deployment.DeploymentName)
```

Comment:
No, the deployment name should not be modified. Escaping is only to construct a workflow id.

Also, updating the internal state this way isn't idempotent.. this will break if you call RegisterTaskQueueWorker twice

Reply:
makes sense to the first point. I modified it to be called only on the generated workflowID and then escaping it there. This shall make the workflowID injective.

Reply:
Nice, so this way we don't ban any char.

Reply:
It doesn't look like you addressed either of these comments? Did you send that reply too soon?

Reply:
sorry, yes addressing them - 

although curious - how does modifying `DeploymentName`/`BuildID` sent by the customer break things and cause idempotentcy issues? We are creating a new DeploymentClient everytime we want to start/update a Deployment workflow right...?